jax/core.py

```diff
@@ -91,6 +96,18 @@ def pretty_print(self, *, source_info=False, print_shapes=True,
   def _repr_pretty_(self, p, cycle):
     return p.text(self.pretty_print(use_color=True))
 
+  def replace(self, *, constvars=None, invars=None, outvars=None, eqns=None,
+              effects=None):
+    constvars = constvars or self.constvars
+    invars = invars or self.invars
+    outvars = outvars or self.outvars
+    eqns = eqns or self.eqns
+    effects = effects or self.effects
+    return Jaxpr(constvars=constvars, invars=invars, outvars=outvars, eqns=eqns,
+                 effects=effects)
+
+def join_effects(effects: Iterable[Effects]) -> Effects:
+  return functools.reduce(lambda a, b: a | b, effects, set())
```

Comment:
```suggestion
  return set.union(*effects)
```

(Though TBH I kind of like not introducing a new abstraction in core.py and instead just calling `set.union` where `core.join_effects` is currently called. Up to you!)

Reply:
Done.